06-release-engineering-analytics.Rmd

```diff
@@ -321,27 +324,116 @@ iteratively and planned for four weekly milestones.
 
 ## Answers
 
-<!-- TODO -->
+In this section, we will give an answer to each of the research questions
+that we have presented in [Section 7.1.1](#research-questions-1).
 
-### RQ1: ...
+### RQ1: Modern Release Engineering Practices
 
-<!-- TODO -->
+This section will give an answer to the question:
+_How is modern release engineering done in practice?_
+
+@adams2016a and @karvonen2017a have described release engineering
+practices that are currently in use in the industry.
+@adams2016a focused on modern release engineering,
+while @karvonen2017a investigated agile release engineering,
+which is a subset of all modern release engineering.
+Both studies agreed that modern release engineering consists
+of the following components:
+
+- **Continuous Integration (CI).**
+
+    To quote @adams2016a: CI "refers to the activity of continuously polling the
+    [version control system] for new commits or merges,
+    checking these revisions out on dedicated build machines,
+    compiling them and running an initial set of tests to check for regressions."
+
+- **Continuous Deployment or Continuous Delivery (CD).**
+
+    These two terms describe the ability to deploy (or deliver)
+    a software product to the end user very frequently.
+    A term related to this is **Rapid Releases (RR)**.
+
+Besides these two components, @adams2016a also identified three other concepts
+that are used in modern release engineering,
+specifically to release software as often as possible and enable CD:
+
+- **Branching and Merging.**
+
+    There are several possible strategies of branching and merging.
+    Typically, in order to enable CD, merging must be done as often as possible.
+
+- **Build System.**
 
-### RQ2: ...
+    Building the software must be done in a consistent way,
+    such that each build produces the same result.
+    With the build configuration stored inside the project,
+    every developer (or automated tool) only needs to issue a single command
+    in order to build the project,
+    instead of manually having to configure the build process every time.
+
+- **Infrastructure-as-Code.**
+
+    In the same alley of "storing configuration",
+    infrastructure-as-code means that the server (or virtual machine)
+    on which the software product is running,
+    can also be automatically configured with code,
+    instead of having to configure each server manually.
+
```

Comment:
For example the `@adams2016a` paper also talks about practices like A/B-testing, blue/green-testing, feature toggles/flags, dark launching, canary deployment... Although not much literature might exists about these practices - especially in our database -, I think these release strategies are worth mentioning because they give a good idea of what is been done in practice.

Reply:
Good point, I only looked at the section headers as I thought that listing all sub-practices would take up too much space. So while I do agree with you, I'm not sure whether we have enough room to list them all :stuck_out_tongue:

Reply:
I agree that it would be a good idea to try to mention some of those, but only if there's space yeah